jlm/llvm/ir/operators/lambda.hpp

```diff
@@ -133,51 +129,52 @@ class node final : public jlm::rvsdg::structural_node
 public:
   class CallSummary;
 
-private:
-  class cviterator;
-  class cvconstiterator;
-
-  class fctargiterator;
-  class fctargconstiterator;
-
-  class fctresiterator;
-  class fctresconstiterator;
-
-  using fctargument_range = jlm::util::iterator_range<fctargiterator>;
-  using fctargument_constrange = jlm::util::iterator_range<fctargconstiterator>;
-
-  using ctxvar_range = jlm::util::iterator_range<cviterator>;
-  using ctxvar_constrange = jlm::util::iterator_range<cvconstiterator>;
-
-  using fctresult_range = jlm::util::iterator_range<fctresiterator>;
-  using fctresult_constrange = jlm::util::iterator_range<fctresconstiterator>;
-
-public:
   ~node() override;
 
 private:
-  node(rvsdg::Region * parent, lambda::operation && op)
-      : structural_node(op, parent, 1)
-  {}
+  node(rvsdg::Region * parent, lambda::operation && op);
 
 public:
-  [[nodiscard]] fctargument_range
-  fctarguments();
-
-  [[nodiscard]] fctargument_constrange
-  fctarguments() const;
-
-  ctxvar_range
-  ctxvars();
-
-  [[nodiscard]] ctxvar_constrange
-  ctxvars() const;
+  /**
+   * \brief Bound context variable
+   *
+   * Context variables may be bound at the point of creation of a
+   * lambda abstraction. These are represented as inputs to the
+   * lambda node itself, and made accessible to the body of the
+   * lambda in the form of an initial argument to the subregion.
+   */
+  struct ContextVar
+  {
+    /**
+     * \brief Input variable bound into lambda node
+     *
+     * The input port into the lambda node that supplies the value
+     * of the context variable bound into the lambda at the
+     * time the lambda abstraction is built.
+     */
+    rvsdg::input * input;
+
+    /**
+     * \brief Access to bound object in subregion.
+     *
+     * Supplies access to the value bound into the lambda abstraction
+     * from inside the region contained in the lambda node. This
+     * evaluates to the value bound into the lambda.
+     */
+    rvsdg::output * inner;
```

Comment:
I originally had that, but actually ran into several inconveniences. I think it is better to consistently return a single type:

1. Callers cannot "do" anything with this information if they are given anything more specific. The primary utility of any def site is that it can be connected to a use site, and for that nothing else is needed. However, if caller wants to do any tracing through the graph, RegionArgument is insufficient and they need to know what exactly it is that is connected and how. So, in practice, RegionArgument does not gain anything.

2. Sites expecting lists of def sites usually expect std::vector<rvsdg::output *>. There are several places in code that need iteration simply to convert from "collection of RegionArgument *" to "collection of rvsdg::output *" as a consequence of being overly specific.

3. Long-term I am not even certain that keeping output / node_output / RegionArgument as class hierarchy makes too much sense. What they are is effectively:

```
struct NodeInput { Node * owner; size_t index; };
struct RegionArgument { Region * owner; size_t index; };
```

One can have `Node` basetype with virtual method `std::variant<Node *, Region *> GetOwner()` (or dynamic_cast or anything like that), but in effect it is mostly wasted overhead in dynamic dispatch, and it is both more memory and time-efficient to just declare it an alternative of the two to begin with. (Besides codifying the fact that this is also the exhaustive list of options).

Reply:
All right, let's leave it like that then.